guidance/rust_code/src/graph/mod.rs

```diff
@@ -1,402 +1,282 @@
 use std::fs::File;
-use std::io::{prelude::*, BufReader, Seek, SeekFrom};
-use geojson::{Feature, FeatureCollection, Value, Geometry, feature::Id};
+use std::io::BufReader;
+use std::f64;
+use std::ops::Sub;
 
-#[derive(PartialEq, Debug)]
-pub(self) struct GPSPoint {
-    latitude: f64,
-    longitude: f64,
-}
-
-#[derive(Debug)]
-pub struct Edge {
-    // For debugging, finding out what line we are looking at on the map
-    name: String,
-    gps_points: Vec<GPSPoint>,
-}
+pub mod conversions;
 
-#[derive(Debug)]
-pub struct Vertex {
-    // Will be used to display key locations to UI
-    name: String,
+use conversions::{IntoTangential};
+use gpx;
+use nalgebra::DMatrix;
 
-    // Will be used to identify adjacent nodes and edges
-    pub(self) gps_point: GPSPoint,
-
-    // Key data to determine the shortest path using Dijkstra's Algorithm
-    parent_vertex: Option<&'static Vertex>,
-    tentative_distance: Option<f64>,
+// Clone trait only for unit testing
+#[derive(Debug, Clone, PartialEq)]
+pub struct Point {
+    pub gps: GPSPointDeg,
+    pub tangential: TangentialPoint
 }
 
-pub struct Graph {
-    pub vertices: Vec<Vertex>,
-    pub edges: Vec<Edge>,
-    pub connection_matrix: Vec<Vec<Option<usize>>> //Matrix of indices of edges
+// Clone Trait only for unit testing
+#[derive(Debug, PartialEq, Clone)]
+pub struct TangentialPoint {
+    pub x: f64,
+    pub y: f64,
+    pub z: f64
 }
 
-impl Edge {
-    //fn distance() -> f64;
-}
+impl Sub for &TangentialPoint {
+    type Output = (f64, f64, f64);
 
-impl Vertex {
-    fn new(name: String, gps_point: GPSPoint) -> Vertex {
-        Vertex {
-            name,
-            gps_point,
-            parent_vertex: None,
-            tentative_distance: None,
-        }
+    fn sub(self, other: Self) -> (f64, f64, f64) {
+        (self.x-other.x, self.y-other.y, self.z-other.z)
     }
 }
 
-impl PartialEq for Vertex {
-    fn eq(&self, other: &Vertex) -> bool {
-        if self.name == other.name && self.gps_point == other.gps_point {
-            return true;
-        }
+impl TangentialPoint {
+    pub fn distance(&self, other: &Self) -> f64 {
+        let (x, y, z) = other - self;
 
-        false
+        (x.powi(2) + y.powi(2) + z.powi(2)).sqrt()
     }
 }
 
-/// Returns (number_of_edges, number_of_vertices) in the Buffer of KML contents
-pub(self) fn number_of_edges_and_vertices_from_buffer(reader: &mut BufReader<File>) -> (u32, u32) {
-    reader.seek(SeekFrom::Start(0)).unwrap();
+// Clone trait only for unit testing
+#[derive(Debug, Clone)]
+pub struct GPSPointDeg {
+    pub lat: f64,
+    pub long: f64,
+    pub height: f64
+}
 
-    let mut number_of_edges = 0;
-    let mut number_of_vertices = 0;
+impl PartialEq for GPSPointDeg {
```

Comment:
Should points be equivalent if they don't have the same height?

Reply:
Unfortunately yes. This is so we can do a comparison between a vertex and a start or end of the edge. The heights are not the same since the height is measured depending on the place marker that was set in google maps.

Reply:
Edge place markers do not have the same hieght as vertex placemarkers unfortunately.